src/ErrorProne.NET/ErrorProne.NET/Rules/ExceptionHandling/ArgumentExceptionUsesStringAnalyzer.cs

```diff
@@ -0,0 +1,88 @@
+        private const string MethodHasNoSuchParamNameRuleDescription = "ArgumentException indicates that a method on the top of the call stack was called with an invalid argument, so it is confusing if ArgumentException.ParamName provides missing paramater name. Consider using nameof operator to avoid typos and post-refactoring errors.";
+
+        private const string ParamNameShouldNotBeStringRuleTitle = "Use nameof instead of string.";
+        private const string ParamNameShouldNotBeStringRuleMessageFormat = "ArgumentException constructor obtains parameter name from string \"{0}\" instead of nameof({0}).";
+        private const string ParamNameShouldNotBeStringRuleDescription = "It is better to obtain string name of the paramater by using nameof operator to avoid typos and post-refactoring errors.";
+
+        private const string Category = "CodeSmell";
+
+        private static readonly DiagnosticDescriptor ParamNameRequiredRule =
+            new DiagnosticDescriptor(RuleIds.ArgumentExceptionParamNameRequired, ParamNameRequiredTitle, ParamNameRequiredMessage, Category, DiagnosticSeverity.Warning, description: ParamNameRequiredDescription, isEnabledByDefault: true);
+        private static readonly DiagnosticDescriptor MethodHasNoSuchParamNameRule =
+            new DiagnosticDescriptor(RuleIds.ArgumentExceptionMethodHasNoSuchParamName, MethodHasNoSuchParamNameRuleTitle, MethodHasNoSuchParamNameRuleMessageFormat, Category, DiagnosticSeverity.Warning, description: MethodHasNoSuchParamNameRuleDescription, isEnabledByDefault: true);
+        private static readonly DiagnosticDescriptor ParamNameShouldNotBeStringRule =
+            new DiagnosticDescriptor(RuleIds.ArgumentExceptionParamNameShouldNotBeString, ParamNameShouldNotBeStringRuleTitle, ParamNameShouldNotBeStringRuleMessageFormat, Category, DiagnosticSeverity.Warning, description: ParamNameShouldNotBeStringRuleDescription, isEnabledByDefault: true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(ParamNameRequiredRule, MethodHasNoSuchParamNameRule, ParamNameShouldNotBeStringRule);
+
+        public override void Initialize(AnalysisContext context) {
+            context.RegisterSyntaxNodeAction(AnalyzeObjectCreation, SyntaxKind.ObjectCreationExpression);
+        }
+
+        private void AnalyzeObjectCreation(SyntaxNodeAnalysisContext context) {
+            var objectCreation = (ObjectCreationExpressionSyntax)context.Node;
+            var method = context.GetCtorSymbol(objectCreation);
+            if (!Equals(method.ContainingType, context.GetClrType<ArgumentException>())) {
+                return;
+            }
+
+            var stringType = context.GetClrType<string>();
+            var paramNameArgument = method.Parameters.FirstOrDefault(p => p.Name == "paramName");
+            if (paramNameArgument == null || !Equals(paramNameArgument.Type, stringType)) {
```

Comment:
The same, imo `paramNameArgument.Type == stringType` is easier to deal with.

BTW, you can simplify this code even more: `paramNameArgument?.Type == stringType`